opentelemetry-sdk/src/logs/growable_array.rs

```diff
@@ -0,0 +1,323 @@
+impl<'a, T: Default + Clone + PartialEq + 'a, const INITIAL_CAPACITY: usize> IntoIterator
+    for &'a GrowableArray<T, INITIAL_CAPACITY>
+{
+    type Item = &'a T;
+    type IntoIter = GrowableArrayIter<'a, T, INITIAL_CAPACITY>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        if self.additional.is_none() || self.additional.as_ref().unwrap().is_empty() {
+            GrowableArrayIter::StackOnly {
+                iter: self.initial.iter().take(self.count),
+            }
+        } else {
+            GrowableArrayIter::Mixed {
+                stack_iter: self.initial.iter().take(self.count),
+                heap_iter: self.additional.as_ref().unwrap().iter(),
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+/// Iterator for referencing elements in a `GrowableArray`.
+///
+/// This iterator allows for iterating over elements in a `GrowableArray` by reference.
+/// It first yields references to the elements stored in the internal array (`initial`),
+/// followed by references to the elements in the vector (`additional`) if present.
+///
+/// # Examples
+///
+pub(crate) enum GrowableArrayIter<'a, T: Default, const INITIAL_CAPACITY: usize> {
+    /// stackonly
+    StackOnly {
+        /// iter
+        iter: std::iter::Take<std::slice::Iter<'a, T>>,
+    },
+    /// hybrid
+    Mixed {
+        /// stack_iter
+        stack_iter: std::iter::Take<std::slice::Iter<'a, T>>,
+        /// heap_iter
+        heap_iter: std::slice::Iter<'a, T>,
+    },
+}
+
+impl<'a, T: Default + Clone, const INITIAL_CAPACITY: usize> Iterator
+    for GrowableArrayIter<'a, T, INITIAL_CAPACITY>
+{
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            GrowableArrayIter::StackOnly { iter } => iter.next(),
+            GrowableArrayIter::Mixed {
+                stack_iter,
+                heap_iter,
+            } => stack_iter.next().or_else(|| heap_iter.next()),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::logs::growable_array::GrowableArray;
+    use opentelemetry::logs::AnyValue;
+    use opentelemetry::Key;
+
+    /// A struct to hold a key-value pair and implement `Default`.
+    #[derive(Clone, Debug, PartialEq)]
+    struct KeyValuePair(Key, AnyValue);
+
+    impl Default for KeyValuePair {
+        fn default() -> Self {
+            KeyValuePair(Key::from_static_str(""), AnyValue::String("".into()))
+        }
+    }
+
+    #[test]
+    fn test_push_and_get() {
+        let mut collection = GrowableArray::<i32>::new();
+        for i in 0..15 {
+            collection.push(i);
+        }
+        for i in 0..15 {
+            assert_eq!(collection.get(i), Some(&(i as i32)));
+        }
+    }
+
+    #[test]
+    fn test_len() {
+        let mut collection = GrowableArray::<i32>::new();
+        for i in 0..15 {
+            collection.push(i);
+        }
+        assert_eq!(collection.len(), 15);
+    }
+
+    #[test]
+    fn test_into_iter() {
```

Comment:
It would be nice to test this with different lengths each representing a unique state of the `GrowableArray`:

1. Empty attributes
2. When the number of attributes is less than the max stack capacity
3. When the number of attributes exactly matches the max stack capacity
4. When the number of attributes exceeds stack capacity but not the initial vec capacity
5. When the number of attributes exceeds stack capacity and also the initial vec capacity

Reply:
Thanks, added tests.